src/dispatch.c

```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+
+        while (((msg = d->queue.head) == NULL) && !d->shutdown)
+            natsCondition_Wait(d->cond, d->mu);
+
+        // Break out only when list is empty
+        if ((msg == NULL) && d->shutdown)
+        {
+            break;
+        }
+
+        _removeFromDispatchQueue(d, msg);
+
+        // Get subscription reference from message and capture values we need
+        // while under lock.
+        natsSubscription *sub = msg->sub;
+        natsConnection *nc = sub->conn;
+        jsSub *jsi = sub->jsi;
+        jsFetch *fetch = (jsi != NULL) ? jsi->fetch : NULL;
+        natsMsgHandler messageCB = sub->msgCb;
+        void *messageClosure = sub->msgCbClosure;
+        natsOnCompleteCB completeCB = sub->onCompleteCB;
+        void *completeCBClosure = sub->onCompleteCBClosure;
+        natsSubscriptionControlMessages *ctrl = sub->control;
+        bool closed = sub->closed;
+        bool draining = sub->draining;
+
+        fetchStatus = _preProcessUserMessage(
+            sub, jsi, fetch, msg,
+            &userMsg, &overLimit, &lastMessageInSub, &lastMessageInFetch, &fcReply);
+
+        // Check the timeout timer.
+        timerNeedReset = false;
+        if (userMsg || (msg == sub->control->sub.timeout))
+        {
+            sub->timeoutSuspended = true;
+            // Need to reset the timer after the user callback returns, but only
+            // if we are already in a timeout, or there are no more messages in
+            // the queue.
+            if (!sub->draining && !sub->closed && (sub->timeout > 0))
+                if (timeout || (sub->ownDispatcher.queue.msgs == 0))
+                {
+                    timerNeedReset = true; // after the callbacks return
+                }
+        }
+
+        // Process synthetic messages
+        if (msg == ctrl->sub.drain)
+        {
+            // Subscription is draining, we are past the last message,
+            // remove the subscription. This will schedule another
+            // control message for the close.
+            nats_unlockDispatcher(d);
+            natsSub_setDrainCompleteState(sub);
+            natsConn_removeSubscription(nc, sub);
+            nats_lockDispatcher(d);
+            continue;
+        }
+        else if (msg == ctrl->sub.close)
+        {
+            nats_unlockDispatcher(d);
+            // Call this in case the subscription was draining.
+            natsSub_setDrainCompleteState(sub);
+
+            if (completeCB != NULL)
+                (*completeCB)(completeCBClosure);
+
+            // Subscription closed, just release
+            natsSub_release(sub);
+
+            nats_lockDispatcher(d);
+            continue;
+        }
+        else if (msg == ctrl->sub.timeout)
+        {
+            nats_unlockDispatcher(d);
+
+            // Invoke the callback with a NULL message.
+            (*messageCB)(nc, sub, NULL, messageClosure);
+
+            nats_lockDispatcher(d);
+
+            if (!draining && !closed)
```

Comment:
We have released/reacquire the lock, but these are evaluated before. The situation may have changed, but I guess the state will be discovered at the next iteration(s) possibly?
The question is: is the subscription protected in case it is closed while the dispatcher lock is released?

Reply:
The sub timeout is a rare event, so 0/5 I should lock the sub and re-acquire these values?

Reply:
Removed the cached variables since the checking is already under the lock.